### ISSUE TITLE
Incorrect __builtin_tzcntll optimization

### ISSUE THREAD

**Issue by zielaj** [llvm/llvm-project#53330]
When you compile this code with clang 13 with `-O3 -march=x86-64 --std=c++2a`

```
#include <bit>
#include <cstdint>

int clipped(std::uint64_t a) {
  return __builtin_ctzll(a) & 0x7f;
}
```

you get this (see [godbolt](https://godbolt.org/z/8YTjE5rWx))

```
bsf rax, rdi
ret
```

This means that `clipped(0)` can return any value, not necessarily in the 0-127 range. This seems to be a bug, as the source code clearly stipulates that the return value should be in the 0-127 range.

GCC documentation, which I'm assuming clang follows, says the following about the the `__builtin_ctz` family:

> Returns the number of trailing 0-bits in x, starting at the least significant bit position. If x is 0, the result is undefined.

This seems to mean that while the result of `__builtin_ctzll(0)` is not specified, it is not undefined behavior in the C++ sense, and the compiler should still guarantee that the result of `clipped(0)` is always in the 0-127 range.

The ideal fix for this bug would fix the generated code for `-march=x86-64` to something like
```
bsf rax, rdi
and rax, 0x7f
ret
```

but on architectures with TZCNT support, eg `-march=skylake` would still optimize out the `and` instruction:
```
tzcnt   rax, rdi
ret
```

**Comment by phoebewang** [comment]
The `and` is removed by [InstCombinePass ](https://godbolt.org/z/9bYooov63)
Maybe we can check if the second parameter of `llvm.cttz.i64` is true before removing the `and`? I'm not sure. AFAIU, undef can be propagated until there is a freeze instruction. I don't know if we should freeze it according to the language semantics.

**Comment by RKSimon** [comment]
CC @rotateright @nunoplopes @nikic

**Comment by LebedevRI** [comment]
This is not a miscompilation given our _current_ langref:
https://alive2.llvm.org/ce/z/_M4veK
https://alive2.llvm.org/ce/z/gr2_Wb

As per https://llvm.org/docs/LangRef.html#llvm-cttz-intrinsic,
> The second argument must be a constant and is a flag to indicate whether the intrinsic should ensure that a zero as the first argument produces a defined result. Historically some architectures did not provide a defined result for zero values as efficiently, and many algorithms are now predicated on avoiding zero-value inputs.

I suppose the question is, even if `undef`, do we have a guarantee that the value is within `[0, bitwidth)`?
I would guess not.

CC @nunoplopes @aqjune

**Comment by nunoplopes** [comment]
> I suppose the question is, even if `undef`, do we have a guarantee that the value is within `[0, bitwidth)`? I would guess not.

No, we can't do that because the hardware doesn't give us that guarantee. See the manual for x86's bsf:
```
IF SRC = 0
    THEN
        ZF ← 1;
        DEST is undefined;
...
```
https://www.felixcloutier.com/x86/bsf

**Comment by nunoplopes** [comment]
TL;DR: this is not a bug. The program executes UB when a=0, so LLVM can do whatever it wants. When a!=0, it returns the correct result.

**Comment by nikic** [comment]
Agree with @nunoplopes, though we should update LangRef to specify that the intrinsic returns poison, not undef, as I believe we started assuming that at some point.

**Comment by nunoplopes** [comment]
> Agree with @nunoplopes, though we should update LangRef to specify that the intrinsic returns poison, not undef, as I believe we started assuming that at some point.

Yep, that's a good idea! We need to get rid of all those undefs.

**Comment by RKSimon** [comment]
@zielaj This might be your best option:
```
#include <bit>
#include <cstdint>

int clipped2(std::uint64_t a) {
  return (a == 0) ? 0x7f : __builtin_ctzll(a) & 0x7f;
}

clipped2(unsigned long):
        bsf     rcx, rdi
        mov     eax, 127
        cmovne  eax, ecx
        ret
        
clipped2(unsigned long):
        tzcnt   rcx, rdi
        mov     eax, 127
        cmovae  eax, ecx
        ret        
```

**Comment by rotateright** [comment]
Proposal to update the LangRef, constant folding, and other bits:
https://reviews.llvm.org/D117912
